### ISSUE TITLE
Data Fixtures

### ISSUE THREAD

**Issue by fridolin-koch** [typeorm/typeorm#1550]
**Issue type:**

[x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[x] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `websql`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

--

I've checked the docs but was not able to find anything related to data fixtures. I've used doctrine/symfony for some years and they have the https://github.com/doctrine/data-fixtures package in their ecosystem. Is there something similar for typeorm? This would be quite useful e.g. for populating the database with defaults or test data.

Thanks!

**Comment by pleerock** [comment]
Simply create a migration and write database population logic there. Use packages like [faker](https://www.npmjs.com/package/faker) to create a fake data. Its basically the same as data fixtures you are asking about.

**Comment by fridolin-koch** [comment]
All right thanks for the answer, I'll try that.

**Comment by ant45de** [comment]
I don't think fixtures should be part of migrations.
Is there currently any other official way to create fixtures?

**Comment by pleerock** [comment]
why they can't?

**Comment by bitwit** [comment]
I was recently trying to solve this and came up with a pretty simple solution.

First i made a `loadFixtures` function: 
```typescript
import { Connection, Repository } from "typeorm"
import * as yaml from 'js-yaml'
import * as fs from 'fs'

export async function loadFixtures(name: string, dbConnection: Connection): Promise<any> {

    let items: any[] = []
    try {
        const file: any = yaml.safeLoad(fs.readFileSync(`./test/fixtures/${name}.yml`, 'utf8'));
        console.log(file)
        items = file["fixtures"]
    } catch (e) {
        console.log("fixtures error", e);
    }

    if(!items) {
        return
    }

    items.forEach(async (item: any) => {
        const entityName = Object.keys(item)[0]
        const data = item[entityName]
        const repo = dbConnection.getRepository(entityName) 
        const newEntity = repo.create(data)
        await repo.createQueryBuilder().insert().into(entityName).values(data).execute()
    })
}
```

And I can load yaml files like this now:

```yaml
fixtures:
    - User:
        id: 1
        email: t@t1.com
        password: "$2a$12$AeLkrNG2Lje156hKYft/CurW8xm/2gGq2Z26OKOwpeIn3XabfM3.." # 12345678
    - User:
        id: 2
        email: second@user.com
        password: "$2a$12$AeLkrNG2Lje156hKYft/CurW8xm/2gGq2Z26OKOwpeIn3XabfM3.." # 12345678
```

I tried using repos directly instead of query builder but ran into some strange issues with duplicate insertions. Once i switched the query builder the problem went away. Not sure if you see any potential pitfalls to a solution like this @pleerock but I just wanted to shared anyway for those who may not want to use migrations.

Personally, I agree it's nicer to keep them out. My fixtures exist in the `test/` folder just as data files. I like yaml for minimal syntax and inline commenting.

TypeORM is great to work with so far, BTW!

**Comment by bitwit** [comment]
I updated the code posted above for some issues I noticed with it plus wanted to note a few things I came across using this technique:

1. I had a problem using `@PrimaryGeneratedColumn("uuid")`, but regular `@PrimaryGeneratedColumn()` was fine. So for uuid's I switched to `@PrimaryColumn` and generated them myself in the constructor.
2. When referring to other tables continue to pass the data as an object: i.e.
```
fixtures:
    - User:
        id: 1
        email: t@t1.com
        password: "$2a$12$AeLkrNG2Lje156hKYft/CurW8xm/2gGq2Z26OKOwpeIn3XabfM3.." # 12345678
        profile: 
          id: 1
```
In this case there is a `User` related to a `Profile` object and you are trying to set the `profileId` column on the `User`. Using `profileId` directly will fail. 
3. Just a reinforced reminder you aren't setting the columns directly through querybuilder. It got me around headaches that `.save()` was causing but it is not raw access to the database.

**Comment by eliranmal** [comment]
@bitwit, i strongly feel you you should wrap all of this into an npm package :shipit:

**Comment by bitwit** [comment]
@eliranmal Thanks! I actually had no clue that anyone had found this post useful until you @ me so that’s great to see.

As far as open sourcing goes the only thing I can point to is this template I made for typescript-express servers. It includes everything I mentioned here and more. Not a standalone lib but more reference to anyone who might find it useful 
https://github.com/bitwit/typescript-express-api-template

**Comment by eliranmal** [comment]
@bitwit oh yes, and there's also [this dude on SO][1], so others probably need this as well :)

the template looks nice, i like it that swagger and test boilerplate are built it.


[1]: https://stackoverflow.com/questions/51198817/typeorm-how-to-seed-database

**Comment by nomadoda** [comment]
@bitwit Any suggestions on how I could seed data with relations?

**Comment by bitwit** [comment]
@nomadoda Can you provide me  more information about what you are trying to do? Does my point #2 on this comment help?: https://github.com/typeorm/typeorm/issues/1550#issuecomment-384160784

**Comment by nomadoda** [comment]
@bitwit I'm trying to follow your example, but storing the data in JS-objects instead of .yaml-files.

I think this pretty accurately describes what I'm trying to do, however keep in mind I'm new to TS and TypeOrm.

*in* ***db/posts.seed.ts:***
```
export const posts = [
  {
    id: uuid(),
    name: 'First post name',
    comments: [
      {
        id: uuid(),
        message: 'First comment',
      },
      {
        id: uuid(),
        message: 'Second comment',
      },
    ]
  },
];
```

*in* ***entity/Post.ts:***
```
@Entity()
export class Post extends BaseEntity {
  @PrimaryColumn('uuid')
  id: string;

  @Column('varchar', { length: 255 })
  name: string;

  @OneToMany(() => Comment, comment => comment.post)
  comments?: Comment[];

  constructor() {
    super();
    this.id = uuid();
  }
}
```

*in* ***entity/Comment.ts:***
```
@Entity()
export class Comment extends BaseEntity {
  @PrimaryColumn('uuid')
  id: string;

  @Column('text')
  message: string;

  @ManyToOne(() => Post, post => post.comments)
  post?: Post;

  constructor() {
    super();
    this.id = uuid();
  }
}
```

*in* ***db/seeder.ts:***
```
import { posts } from './posts.seed.ts';

export const seedPosts = async () => {
  const dbConnection = await createTypeormConnection();
  const promises = posts.map((post: any) => {
    return dbConnection.createQueryBuilder().insert().into('Post').values(post).execute();
  });
  await Promise.all(promises); // or in my case I run them sequentially by chaining the promises
};
```

Running `seedPosts` results in a new post, but no comments are added.

**Comment by bitwit** [comment]
Ah, gotcha. I did not test this with any kind of *-to-many relationship yet or trying to populate anything more that the 'profileId' column. I'm not working on typeorm much these days so I can't offer to much assistance testing this right now but my thoughts in some first things to try:

It probably doesn't work because in the database you wouldn't be storing your array of comments directly on that entity. Under the hood there is some post-comments table making the associations and that's the one you need to populate, in addition to your comments table. I'm not familiar with how typeORM is setting up these relationships or what tables it's generating for you but there's probably a way to gain control of this.

So try flattening out what you are trying to import and putting the appropriate things in the relevant 3 tables of the relationship i.e. `Post`, `Comment`, and the currently implicit `PostComment` relationship Entity

**Comment by caseyduquettesc** [comment]
For anyone using @bitwit's nice fixture loader, if you're seeing foreign key constraint failures or you're selecting data immediately after the fixtures and randomly seeing null results, you may want to tweak the `for` loop. As-is, the inserts for each entity type get run in parallel, meaning some foreign entities may not exist when you expect them to, or your first test may not return any data.

The fix is pretty simple:
```ts
for (const item of items) {
  const entityName = Object.keys(item)[0];
  const data = item[entityName];
  await dbConnection
    .createQueryBuilder()
    .insert()
    .into(entityName)
    .values(data)
    .execute();
  }
```

**Comment by feimosi** [comment]
For anyone who also encountered foreign key constraint errors and would like to insert values with TypeORM relation cascading: 
Apparently `createQueryBuilder().insert()` doesn't do cascade inserts so use `getRepository.save()` instead.

```ts
  for (const item of items) {
    const entityName = Object.keys(item)[0];
    const data = item[entityName];

    await getRepository(entityName)
      .save(data);
  }
```

**Comment by abcd-ca** [comment]
In my case I solved it within my reset script which is only called in a global `beforeEach`. My steps are to,
1. init db (creates a default db connection with `await createConnection()`
2. drop all known tables 
```
await connection.manager.query(`DROP TABLE IF EXISTS ${tableNamesCSV} CASCADE`);
```
3. close the connection, 
```
  const connection = getConnection();
  await connection.close();
  // creates the tables again
  await createConnection();
```
4. seed the db with new data

Another key was to rely on TypeORM's `getConnection()` function. You can check to see if it doesn't have a connection and conditionally call `await createConnection()` if not – after which `getConnection()` will have a connection.

Another issues I overcame which seem to have been relevant:
* my drop table query needed to not have quotes around the table names when I was dropping one by one. Once that was working I moved to a multi-table drop query (CSV)
* I renamed my `User` entity to `FooUser`. It seemed that Postgres was complaining that when trying to drop the `user` table, `user` is a reserved word in Postgres. Now it drops `foo_user` and is OK.

**Comment by havenchyk** [comment]
@abcd-ca 

> drop all known tables

you could use also `connection.synchronize(true)`, it drops all the data from your tables.

**Comment by abcd-ca** [comment]
@havenchyk Interesting – I wasn't clear on how that is related to the environment variable I already have set, `TYPEORM_SYNCHRONIZE = true` used by `createConnection`.

**Comment by havenchyk** [comment]
Well, @abcd-ca it seems to be used as a config value https://github.com/typeorm/typeorm/blob/7985e2e992842e8a501f315ef44c77b3d188fd64/src/connection/options-reader/ConnectionOptionsEnvReader.ts#L31, but with `connection.synchronize(true)` you can clear all your data per test without creating a new connection every time.

```
    /**
     * Creates database schema for all entities registered in this connection.
     * Can be used only after connection to the database is established.
     *
     * @param dropBeforeSync If set to true then it drops the database with all its tables and data
     */
    async synchronize(dropBeforeSync: boolean = false): Promise<void>
```

https://github.com/typeorm/typeorm/blob/a4dec02cc59d3219a29c7be0322af2253e1452dc/src/connection/Connection.ts#L236

**Comment by havenchyk** [comment]
@abcd-ca if you think that documentation lacks on this, please, create a PR 😄

**Comment by RobinCK** [comment]
My implementation of fixtures: https://github.com/RobinCK/typeorm-fixtures

**Comment by bitwit** [comment]
@RobinCK This looks pretty dope! Well done.

**Comment by kferrone** [comment]
Anyone know if it is possible to catch the table created event if there is one? I would simply like to load `lookup table` data directly after a table has been created or directly after the `synchronize` runs.

**Comment by luisesantos** [comment]
Hi guys, I want to use typeorm-fixtures for my unit/integration testings with SQLite. @RobinCK can I run typeorm-fixtures programmatically ?. I am exploring if running fixtures on "beforeAll" and cleaning on "afterAll" on a SQLite db (:memory) is actually doable. 

Thanks a lot in advance
Luis

Technology stack: Koa, TypeORM, Jest, Supertest

**Comment by RobinCK** [comment]
I think that this will help you https://github.com/RobinCK/typeorm-fixtures/issues/8

**Comment by luisesantos** [comment]
Awesome thanks a lot.

**Comment by luisesantos** [comment]
@RobinCK sorry to bother but would something like that work with fakerJS ?. If so could you point to a better example of how to integrate with it. At the moment when I load the yml files from the fixtures  directory the faker methods are not recognized.

Thanks again in advance
Luis

**Comment by rsp** [comment]
@bitwit a quick comment to your `loadFixtures` function from [this comment](https://github.com/typeorm/typeorm/issues/1550#issuecomment-381039669) above.

This will not wait for `.execute()` to finish before doing next iterations, possibly leading to race conditions:
```
    items.forEach(async (item: any) => {
        const entityName = Object.keys(item)[0]
        const data = item[entityName]
        await dbConnection.createQueryBuilder().insert().into(entityName).values(data).execute()
    })
```

If you want to insert everything in order (it is sometimes needed if there is some referential integrity that can kick in) then you can use:
```
    for (const item of items) {
        const entityName = Object.keys(item)[0]
        const data = item[entityName]
        await dbConnection.createQueryBuilder().insert().into(entityName).values(data).execute()
    }
```
and now the loop will wait with the next iteration before the previous `.execute()` completed.

**Comment by jeanfortheweb** [comment]
I dont know if anyone still cares, but I'm working on a library for this use case for some months now and just released 2.0. Maybe some of you want to take a look:

https://github.com/jeanfortheweb/typeorm-fixture-builder
